File: Samples/Unity/PlayFabLeaderboardsUnity/Assets/PlayFabSdk/Matchmaker/PlayFabMatchmakerAPI.cs
#if ENABLE_PLAYFABSERVER_API
using System;
using System.Collections.Generic;
using PlayFab.MatchmakerModels;
using PlayFab.Internal;
using PlayFab.Json;
using PlayFab.Public;

namespace PlayFab
{
    /// <summary>
    /// Enables the use of an external match-making service in conjunction with PlayFab hosted Game Server instances
    /// </summary>
    public static class PlayFabMatchmakerAPI
    {
        static PlayFabMatchmakerAPI() {}


        /// <summary>
        /// Clear the Client SessionToken which allows this Client to call API calls requiring login.
        /// A new/fresh login will be required after calling this.
        /// </summary>
        public static void ForgetAllCredentials()
        {
            PlayFabHttp.ForgetAllCredentials();
        }

        /// <summary>
        /// Validates a user with the PlayFab service
        /// </summary>
        public static void AuthUser(AuthUserRequest request, Action<AuthUserResponse> resultCallback, Action<PlayFabError> errorCallback, object customData = null, Dictionary<string, string> extraHeaders = null)
        {
            if (PlayFabSettings.DeveloperSecretKey == null) throw new PlayFabException(PlayFabExceptionCode.DeveloperKeyNotSet,"Must have PlayFabSettings.DeveloperSecretKey set to call this method");

            PlayFabHttp.MakeApiCall("/Matchmaker/AuthUser", request, AuthType.DevSecretKey, resultCallback, errorCallback, customData, extraHeaders);
        }

        /// <summary>
        /// Informs the PlayFab game server hosting service that the indicated user has joined the Game Server Instance specified
        /// </summary>
        public static void PlayerJoined(PlayerJoinedRequest request, Action<PlayerJoinedResponse> resultCallback, Action<PlayFabError> errorCallback, object customData = null, Dictionary<string, string> extraHeaders = null)
        {
            if (PlayFabSettings.DeveloperSecretKey == null) throw new PlayFabException(PlayFabExceptionCode.DeveloperKeyNotSet,"Must have PlayFabSettings.DeveloperSecretKey set to call this method");

            PlayFabHttp.MakeApiCall("/Matchmaker/PlayerJoined", request, AuthType.DevSecretKey, resultCallback, errorCallback, customData, extraHeaders);
        }

        /// <summary>
        /// Informs the PlayFab game server hosting service that the indicated user has left the Game Server Instance specified
        /// </summary>
        public static void PlayerLeft(PlayerLeftRequest request, Action<PlayerLeftResponse> resultCallback, Action<PlayFabError> errorCallback, object customData = null, Dictionary<string, string> extraHeaders = null)
        {
            if (PlayFabSettings.DeveloperSecretKey == null) throw new PlayFabException(PlayFabExceptionCode.DeveloperKeyNotSet,"Must have PlayFabSettings.DeveloperSecretKey set to call this method");

            PlayFabHttp.MakeApiCall("/Matchmaker/PlayerLeft", request, AuthType.DevSecretKey, resultCallback, errorCallback, customData, extraHeaders);
        }

        /// <summary>
        /// Instructs the PlayFab game server hosting service to instantiate a new Game Server Instance
        /// </summary>
        public static void StartGame(StartGameRequest request, Action<StartGameResponse> resultCallback, Action<PlayFabError> errorCallback, object customData = null, Dictionary<string, string> extraHeaders = null)
        {
            if (PlayFabSettings.DeveloperSecretKey == null) throw new PlayFabException(PlayFabExceptionCode.DeveloperKeyNotSet,"Must have PlayFabSettings.DeveloperSecretKey set to call this method");

            PlayFabHttp.MakeApiCall("/Matchmaker/StartGame", request, AuthType.DevSecretKey, resultCallback, errorCallback, customData, extraHeaders);
        }

        /// <summary>
        /// Retrieves the relevant details for a specified user, which the external match-making service can then use to compute
        /// effective matches
        /// </summary>
        public static void UserInfo(UserInfoRequest request, Action<UserInfoResponse> resultCallback, Action<PlayFabError> errorCallback, object customData = null, Dictionary<string, string> extraHeaders = null)
        {
            if (PlayFabSettings.DeveloperSecretKey == null) throw new PlayFabException(PlayFabExceptionCode.DeveloperKeyNotSet,"Must have PlayFabSettings.DeveloperSecretKey set to call this method");

            PlayFabHttp.MakeApiCall("/Matchmaker/UserInfo", request, AuthType.DevSecretKey, resultCallback, errorCallback, customData, extraHeaders);
        }


    }
}
#endif

File: Samples/Xbox/Kits/PlayFabSDK/source/PlayFabMatchmakerApi.cpp

#ifdef ENABLE_PLAYFABSERVER_API

#include "playfab/PlayFabMatchmakerApi.h"
#include "playfab/PlayFabHttp.h"
#include "playfab/PlayFabSettings.h"

namespace PlayFab
{
    using namespace MatchmakerModels;

    size_t PlayFabMatchmakerAPI::Update()
    {
        return PlayFabHttp::Get().Update();
    }

    void PlayFabMatchmakerAPI::ForgetAllCredentials()
    {
        return PlayFabSettings::ForgetAllCredentials();
    }

    // PlayFabMatchmaker APIs

    void PlayFabMatchmakerAPI::AuthUser(
        AuthUserRequest& request,
        ProcessApiCallback<AuthUserResponse> callback,
        ErrorCallback errorCallback,
        void* customData
    )
    {

        IPlayFabHttp& http = IPlayFabHttp::Get();
        auto requestJson = request.ToJson();
        http.AddRequest(L"/Matchmaker/AuthUser", L"X-SecretKey", PlayFabSettings::developerSecretKey, requestJson, OnAuthUserResult, SharedVoidPointer((callback == nullptr) ? nullptr : new ProcessApiCallback<AuthUserResponse>(callback)), errorCallback, customData);
    }

    void PlayFabMatchmakerAPI::OnAuthUserResult(CallRequestContainer& request)
    {
        AuthUserResponse outResult;
        outResult.FromJson(request.errorWrapper.Data);
        outResult.Request = request.request;

        auto internalPtr = request.successCallback.get();
        if (internalPtr != nullptr)
        {
            auto callback = (*static_cast<ProcessApiCallback<AuthUserResponse> *>(internalPtr));
            callback(outResult, request.customData);
        }
    }

    void PlayFabMatchmakerAPI::PlayerJoined(
        PlayerJoinedRequest& request,
        ProcessApiCallback<PlayerJoinedResponse> callback,
        ErrorCallback errorCallback,
        void* customData
    )
    {

        IPlayFabHttp& http = IPlayFabHttp::Get();
        auto requestJson = request.ToJson();
        http.AddRequest(L"/Matchmaker/PlayerJoined", L"X-SecretKey", PlayFabSettings::developerSecretKey, requestJson, OnPlayerJoinedResult, SharedVoidPointer((callback == nullptr) ? nullptr : new ProcessApiCallback<PlayerJoinedResponse>(callback)), errorCallback, customData);
    }

    void PlayFabMatchmakerAPI::OnPlayerJoinedResult(CallRequestContainer& request)
    {
        PlayerJoinedResponse outResult;
        outResult.FromJson(request.errorWrapper.Data);
        outResult.Request = request.request;

        auto internalPtr = request.successCallback.get();
        if (internalPtr != nullptr)
        {
            auto callback = (*static_cast<ProcessApiCallback<PlayerJoinedResponse> *>(internalPtr));
            callback(outResult, request.customData);
        }
    }

    void PlayFabMatchmakerAPI::PlayerLeft(
        PlayerLeftRequest& request,
        ProcessApiCallback<PlayerLeftResponse> callback,
        ErrorCallback errorCallback,
        void* customData
    )
    {

        IPlayFabHttp& http = IPlayFabHttp::Get();
        auto requestJson = request.ToJson();
        http.AddRequest(L"/Matchmaker/PlayerLeft", L"X-SecretKey", PlayFabSettings::developerSecretKey, requestJson, OnPlayerLeftResult, SharedVoidPointer((callback == nullptr) ? nullptr : new ProcessApiCallback<PlayerLeftResponse>(callback)), errorCallback, customData);
    }

    void PlayFabMatchmakerAPI::OnPlayerLeftResult(CallRequestContainer& request)
    {
        PlayerLeftResponse outResult;
        outResult.FromJson(request.errorWrapper.Data);
        outResult.Request = request.request;

        auto internalPtr = request.successCallback.get();
        if (internalPtr != nullptr)
        {
            auto callback = (*static_cast<ProcessApiCallback<PlayerLeftResponse> *>(internalPtr));
            callback(outResult, request.customData);
        }
    }

    void PlayFabMatchmakerAPI::StartGame(
        StartGameRequest& request,
        ProcessApiCallback<StartGameResponse> callback,
        ErrorCallback errorCallback,
        void* customData
    )
    {

        IPlayFabHttp& http = IPlayFabHttp::Get();
        auto requestJson = request.ToJson();
        http.AddRequest(L"/Matchmaker/StartGame", L"X-SecretKey", PlayFabSettings::developerSecretKey, requestJson, OnStartGameResult, SharedVoidPointer((callback == nullptr) ? nullptr : new ProcessApiCallback<StartGameResponse>(callback)), errorCallback, customData);
    }

    void PlayFabMatchmakerAPI::OnStartGameResult(CallRequestContainer& request)
    {
        StartGameResponse outResult;
        outResult.FromJson(request.errorWrapper.Data);
        outResult.Request = request.request;

        auto internalPtr = request.successCallback.get();
        if (internalPtr != nullptr)
        {
            auto callback = (*static_cast<ProcessApiCallback<StartGameResponse> *>(internalPtr));
            callback(outResult, request.customData);
        }
    }

    void PlayFabMatchmakerAPI::UserInfo(
        UserInfoRequest& request,
        ProcessApiCallback<UserInfoResponse> callback,
        ErrorCallback errorCallback,
        void* customData
    )
    {

        IPlayFabHttp& http = IPlayFabHttp::Get();
        auto requestJson = request.ToJson();
        http.AddRequest(L"/Matchmaker/UserInfo", L"X-SecretKey", PlayFabSettings::developerSecretKey, requestJson, OnUserInfoResult, SharedVoidPointer((callback == nullptr) ? nullptr : new ProcessApiCallback<UserInfoResponse>(callback)), errorCallback, customData);
    }

    void PlayFabMatchmakerAPI::OnUserInfoResult(CallRequestContainer& request)
    {
        UserInfoResponse outResult;
        outResult.FromJson(request.errorWrapper.Data);
        outResult.Request = request.request;

        auto internalPtr = request.successCallback.get();
        if (internalPtr != nullptr)
        {
            auto callback = (*static_cast<ProcessApiCallback<UserInfoResponse> *>(internalPtr));
            callback(outResult, request.customData);
        }
    }
}

#endif
